File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/webrtcclient/General/Prefs.java
package kore.botssdk.audiocodes.webrtcclient.General;

import android.content.Context;
import android.media.AudioManager;

import com.audiocodes.mv.webrtcsdk.session.ACCallStatistics;
import com.audiocodes.mv.webrtcsdk.session.DTMFOptions;

import kore.botssdk.R;
import kore.botssdk.audiocodes.webrtcclient.Structure.SipAccount;


public class Prefs extends BasePrefs {

    private static final String TAG = "Prefs";

    public static void setFirstLogin(Context context, boolean firstLogin) {
        Log.d(TAG, "setFirstLogin: " + firstLogin);
        putBoolean(context, "FIRST_LOGIN", firstLogin);
    }

    public static boolean isFirstLogin(Context context) {
        Log.d(TAG, "get isFirstLogin: " + getBoolean(context, "FIRST_LOGIN", true));
        return getBoolean(context, "FIRST_LOGIN", true);
    }

    public static boolean isAppOpen(Context context) {
        return getBoolean(context, "APP_OPEN");
    }

    public static void setAppOpen(Context context, boolean state) {
        putBoolean(context, "APP_OPEN", state);
    }

    public static void setSipAccount(Context context, SipAccount sipAccount) {
        putString(context, "SipAccountDisplayName", sipAccount.getDisplayName());
        putString(context, "SipAccountUsername", sipAccount.getUsername());
        putString(context, "SipAccountPassword", sipAccount.getPassword());
        putString(context, "SipAccountDomain", sipAccount.getDomain());
        putString(context, "SipAccountProxy", sipAccount.getProxy());
        putInt(context, "SipAccountPort", sipAccount.getPort());
        putString(context, "SipAccountTransport", sipAccount.getTransport().toString());
    }

    public static SipAccount getSipAccount(Context context) {
        SipAccount sipAccount = new SipAccount(context);

        sipAccount.setDisplayName(getString(context, "SipAccountDisplayName"));
        sipAccount.setUsername(getString(context, "SipAccountUsername"));
        sipAccount.setPassword(getString(context, "SipAccountPassword"));
        sipAccount.setDomain(getString(context, "SipAccountDomain"));
        sipAccount.setProxy(getString(context, "SipAccountProxy"));
        sipAccount.setPort(getInt(context, "SipAccountPort"));
        //String tempTransport = getString("SipAccountTransport");
        //sipAccount.setTransport(Transport.valueOf(tempTransport));
        sipAccount.setTransport(AppUtils.getTransport(context, getString(context, "SipAccountTransport")));

        return sipAccount;
    }

    public static void setCallStats(Context context, ACCallStatistics acCallStatistics) {
        putClass(context, "ACCallStats", acCallStatistics);
    }

    public static ACCallStatistics getCallStats(Context context) {
        ACCallStatistics acCallStatistics = (ACCallStatistics) getClass(context, "ACCallStats", ACCallStatistics.class);

        return acCallStatistics;
    }

    public static void setDTMFType(Context context, DTMFOptions.DTMFMethod dtmfType) {
        putClass(context, "DTMFMethod", dtmfType);
    }

    public static DTMFOptions.DTMFMethod getDTMFType(Context context) {
        //the default value is DTMFOptions.DTMFMethod.WEBRTC
        DTMFOptions.DTMFMethod dtmfMethod = (DTMFOptions.DTMFMethod) getClass(context, "DTMFMethod", DTMFOptions.DTMFMethod.class, DTMFOptions.DTMFMethod.WEBRTC);

        return dtmfMethod;
    }

    public static void setLogLevel(Context context, Log.LogLevel logLevel) {
        putClass(context, "LogLevel", logLevel);
    }

    public static Log.LogLevel getLogLevel(Context context) {
        //the default value is in configuration file
        String defaultLogLevel = context.getString(R.string.log_level_default);
        Log.LogLevel logLevel = (Log.LogLevel) getClass(context, "LogLevel", Log.LogLevel.class, Log.LogLevel.valueOf(defaultLogLevel));

        return logLevel;
    }

    public static boolean isAutoRedirect(Context context) {
        return getBoolean(context, "AutoRedirect");
    }

    public static void setAutoRedirect(Context context, boolean state) {
        putBoolean(context, "AutoRedirect", state);
    }

    public static boolean isRedirectCall(Context context) {
        return getBoolean(context, "RedirectCall");
    }

    public static void setRedirectCall(Context context, boolean state) {
        putBoolean(context, "RedirectCall", state);
    }

    public static String getRedirectCallUser(Context context) {
        return getString(context, "RedirectCallUser", context.getString(R.string.default_redirect_call_user));
    }

    public static void setRedirectCallUser(Context context, String redirectCallUser) {
        putString(context, "RedirectCallUser", redirectCallUser);
    }

    public static boolean isVideoHardware(Context context) {
        return getBoolean(context, "VideoHardware");
    }

    public static void setVideoHardware(Context context, boolean state) {
        putBoolean(context, "VideoHardware", state);
    }


    public static boolean getAutoLogin(Context context) {
        return getBoolean(context, "AutoLogin", false);
    }

    public static void setAutoLogin(Context context, boolean enable) {
        putBoolean(context, "AutoLogin", enable);
    }

    public static boolean getDisconnectBrokenConnection(Context context) {
        return getBoolean(context, "DisconnectBrokenConnection", true);
    }

    public static void setDisconnectBrokenConnection(Context context, boolean enable) {
        putBoolean(context, "DisconnectBrokenConnection", enable);
    }

    public static String getSecondCall(Context context) {
        return getString(context, "SecondCall", context.getString(R.string.sip_default_second_call));
    }

    public static void setSecondCall(Context context, String call) {
        putString(context, "SecondCall", call);
    }

    public static String getTransferCall(Context context) {
        return getString(context, "TransferCall", context.getString(R.string.sip_default_transfer_call));
    }

    public static void setTransferCall(Context context, String call) {
        putString(context, "TransferCall", call);
    }

    public static boolean usePush(Context context) {
        return getBoolean(context, "USE_PUSH");
    }

    public static void setUsePush(Context context, boolean usePush) {
        putBoolean(context, "USE_PUSH", usePush);
    }

    //volume control
    public static int getRingVolume(Context context) {
        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        int maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_RING);
        //default value should be max
        int res = getInt(context, "RING_VOLUME", maxVolume);
        Log.d(TAG, "get RING_VOLUME: " + res);
        return res;
    }

    public static void setRingVolume(Context context, int vol) {
        Log.d(TAG, "set RING_VOLUME: " + vol);
        putInt(context, "RING_VOLUME", vol);
    }

    public static int getCallVolume(Context context) {
        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        int maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_VOICE_CALL);
        //default value should be max
        int res = getInt(context, "CALL_VOLUME", maxVolume);
        Log.d(TAG, "get CALL_VOLUME: " + res);
        return res;
    }

    public static void setCallVolume(Context context, int vol) {
        Log.d(TAG, "set CALL_VOLUME: " + vol);
        putInt(context, "CALL_VOLUME", vol);
    }

    public static int getPrevRingVolume(Context context) {
        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        int maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_RING);
        //default value should be max
        int res = getInt(context, "PREV_RING_VOLUME", maxVolume);
        Log.d(TAG, "get PREV_RING_VOLUME: " + res);
        return res;
    }

    public static void setPrevRingVolume(Context context, int vol) {
        Log.d(TAG, "set PREV_RING_VOLUME: " + vol);
        putInt(context, "PREV_RING_VOLUME", vol);
    }

    public static int getPrevCallVolume(Context context) {
        AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
        int maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_VOICE_CALL);
        //default value should be max
        int res = getInt(context, "PREV_CALL_VOLUME", maxVolume);
        Log.d(TAG, "get PREV_CALL_VOLUME: " + res);
        return res;
    }

    public static void setPrevCallVolume(Context context, int vol) {
        Log.d(TAG, "set PREV_CALL_VOLUME: " + vol);
        putInt(context, "PREV_CALL_VOLUME", vol);
    }

//    public static int getRingtoneVolume()
//    {
//        AudioManager audioManager = (AudioManager)context.getSystemService("audio");
//        int maxVolume = audioManager.getStreamMaxVolume(0);
//        //default value should be max
//        getInt("RINGTONE_VOLUME", maxVolume);
//    }
//    public static void SetRingtoneVolume(int vol)
//    {
//        putInt("RINGTONE_VOLUME", vol);
//    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/webrtcclient/general/Prefs.kt
package com.kore.ui.audiocodes.webrtcclient.general

import android.content.Context
import android.media.AudioManager
import com.audiocodes.mv.webrtcsdk.session.ACCallStatistics
import com.audiocodes.mv.webrtcsdk.session.DTMFOptions.DTMFMethod
import com.kore.ui.R
import com.kore.ui.audiocodes.webrtcclient.general.AppUtils.getTransport
import com.kore.ui.audiocodes.webrtcclient.general.Log.d
import com.kore.ui.audiocodes.webrtcclient.structure.SipAccount

object Prefs : BasePrefs() {
    private const val TAG = "Prefs"
    fun getIsFirstLogin(context: Context): Boolean {
        d(TAG, "get isFirstLogin: " + getBoolean(context, "FIRST_LOGIN", true))
        return getBoolean(context, "FIRST_LOGIN", true)
    }

    fun setIsFirstLogin(context: Context, firstLogin: Boolean) {
        d(TAG, "setFirstLogin: $firstLogin")
        putBoolean(context, "FIRST_LOGIN", firstLogin)
    }

    fun getIsAppOpen(context: Context): Boolean = getBoolean(context, "APP_OPEN")
    fun setIsAppOpen(context: Context, state: Boolean) {
        putBoolean(context, "APP_OPEN", state)
    }

    fun getSipAccount(context: Context): SipAccount {
        val sipAccount = SipAccount(context)
        sipAccount.displayName = getString(context, "SipAccountDisplayName")
        sipAccount.username = getString(context, "SipAccountUsername")
        sipAccount.password = getString(context, "SipAccountPassword")
        sipAccount.domain = getString(context, "SipAccountDomain")
        sipAccount.proxy = getString(context, "SipAccountProxy")
        sipAccount.port = getInt(context, "SipAccountPort")
        //String tempTransport = getString(context, "SipAccountTransport");
        //sipAccount.setTransport(Transport.valueOf(tempTransport));
        sipAccount.transport = getTransport(context, getString(context, "SipAccountTransport"))
        return sipAccount
    }

    fun setSipAccount(context: Context, sipAccount: SipAccount) {
        putString(context, "SipAccountDisplayName", sipAccount.displayName)
        putString(context, "SipAccountUsername", sipAccount.username)
        putString(context, "SipAccountPassword", sipAccount.password)
        putString(context, "SipAccountDomain", sipAccount.domain)
        putString(context, "SipAccountProxy", sipAccount.proxy)
        putInt(context, "SipAccountPort", sipAccount.port)
        putString(context, "SipAccountTransport", sipAccount.transport.toString())
    }

    fun getCallStats(context: Context): ACCallStatistics? = getClass(context, "ACCallStats", ACCallStatistics::class.java) as ACCallStatistics?
    fun setCallState(context: Context, acCallStatistics: ACCallStatistics) {
        putClass(context, "ACCallStats", acCallStatistics)
    }

    fun getDTMFType(context: Context) =//the default value is DTMFOptions.DTMFMethod.WEBRTC
        getClass(context, "DTMFMethod", DTMFMethod::class.java, DTMFMethod.WEBRTC) as DTMFMethod?

    fun setDTMFType(context: Context, dtmfType: DTMFMethod) {
        putClass(context, "DTMFMethod", dtmfType)
    }

    fun getLogLevel(context: Context): Log.LogLevel? {
        //the default value is in configuration file
        val defaultLogLevel = context.getString(R.string.log_level_default)
        return getClass(context, "LogLevel", Log.LogLevel::class.java, Log.LogLevel.valueOf(defaultLogLevel)) as Log.LogLevel?
    }

    fun setLogLevel(context: Context, logLevel: Log.LogLevel) {
        putClass(context, "LogLevel", logLevel)
    }

    fun getIsAutoRedirect(context: Context): Boolean = getBoolean(context, "AutoRedirect")
    fun setIsAutoRedirect(context: Context, state: Boolean) {
        putBoolean(context, "AutoRedirect", state)
    }

    fun getIsRedirectCall(context: Context): Boolean = getBoolean(context, "RedirectCall")
    fun setIsRedirectCall(context: Context, state: Boolean) {
        putBoolean(context, "RedirectCall", state)
    }

    fun getRedirectCallUser(context: Context): String? {
        return getString(context, "RedirectCallUser", context.getString(R.string.default_redirect_call_user))
    }

    fun setRedirectCallUser(context: Context, redirectCallUser: String) {
        putString(context, "RedirectCallUser", redirectCallUser)
    }

    fun getIsVideoHardware(context: Context): Boolean = getBoolean(context, "VideoHardware")
    fun setIsVideoHardware(context: Context, state: Boolean) {
        putBoolean(context, "VideoHardware", state)
    }

    fun getAutoLogin(context: Context): Boolean = getBoolean(context, "AutoLogin", false)
    fun setAutoLogin(context: Context, enable: Boolean) {
        putBoolean(context, "AutoLogin", enable)
    }

    fun getDisconnectBrokenConnection(context: Context) = getBoolean(context, "DisconnectBrokenConnection", true)
    fun setDisconnectBrokenConnection(context: Context, enable: Boolean) {
        putBoolean(context, "DisconnectBrokenConnection", enable)
    }

    fun getSecondCall(context: Context): String? {
        return getString(context, "SecondCall", context.getString(R.string.sip_default_second_call))
    }

    fun setSecondCall(context: Context, call: String) {
        putString(context, "SecondCall", call)
    }

    fun getTransferCall(context: Context): String? {
        return getString(context, "TransferCall", context.getString(R.string.sip_default_transfer_call))
    }

    fun setTransferCall(context: Context, call: String) {
        putString(context, "TransferCall", call)
    }

    fun usePush(context: Context): Boolean {
        return getBoolean(context, "USE_PUSH")
    }

    fun setUsePush(context: Context, usePush: Boolean) {
        putBoolean(context, "USE_PUSH", usePush)
    }

    //volume control
    fun getRingVolume(context: Context): Int {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_RING)
        //default value should be max
        val volume = getInt(context, "RING_VOLUME", maxVolume)
        d(TAG, "get RING_VOLUME: $volume")
        return volume
    }

    fun setRingVolume(context: Context, vol: Int) {
        d(TAG, "set RING_VOLUME: $vol")
        putInt(context, "RING_VOLUME", vol)
    }

    fun getCallVolume(context: Context): Int {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_VOICE_CALL)
        //default value should be max
        val volume = getInt(context, "CALL_VOLUME", maxVolume)
        d(TAG, "get CALL_VOLUME: $volume")
        return volume
    }

    fun setCallVolume(context: Context, vol: Int) {
        d(TAG, "set CALL_VOLUME: $vol")
        putInt(context, "CALL_VOLUME", vol)
    }

    private var prevRingVolume: Int = 0
    fun getPrevRingVolume(context: Context): Int {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_RING)
        //default value should be max
        val volume = getInt(context, "PREV_RING_VOLUME", maxVolume)
        d(TAG, "get PREV_RING_VOLUME: $volume")
        return volume
    }

    fun setPrevRingVolume(context: Context, vol: Int) {
        d(TAG, "set PREV_RING_VOLUME: $vol")
        putInt(context, "PREV_RING_VOLUME", vol)
    }

    fun getPrevCallVolume(context: Context): Int {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        val maxVolume = audioManager.getStreamMaxVolume(AudioManager.STREAM_VOICE_CALL)
        //default value should be max
        val volume = getInt(context, "PREV_CALL_VOLUME", maxVolume)
        d(TAG, "get PREV_CALL_VOLUME: $volume")
        return volume
    }

    fun setPrevCallVolume(context: Context, vol: Int) {
        d(TAG, "set PREV_CALL_VOLUME: $vol")
        putInt(context, "PREV_CALL_VOLUME", vol)
    }

//    public static int getRingtoneVolume()
    //    {
    //        AudioManager audioManager = (AudioManager)BotApplication.Companion.context.getSystemService("audio");
    //        int maxVolume = audioManager.getStreamMaxVolume(0);
    //        //default value should be max
    //        getInt("RINGTONE_VOLUME", maxVolume);
    //    }
    //    public static void SetRingtoneVolume(int vol)
    //    {
    //        putInt("RINGTONE_VOLUME", vol);
    //    }
}
